graphistry/compute/symbolic.py

```diff
@@ -0,0 +1,195 @@
+#import sympy
+from typing import TYPE_CHECKING
+from collections import Counter
+import numpy as np
+#from graphistry.ai_utils import build_annoy_index, query_by_vector, DISTANCE
+from graphistry.text_utils import SearchToGraphMixin
+import pandas as pd
+
+try:
+    import symai as ai
+    from symai import *
+except ImportError:
+    ai = None
+    
+__doc__  =  """graphistry.compute symbolic
+                providing symbolic graph operations
+                (e.g., computing the characteristic polynomial of a graph's adjacency matrix)
+                or symbolic graph operations using the graphistry.ai chatbot"""
+
+
+
+if TYPE_CHECKING:
+    MIXIN_BASE = SearchToGraphMixin
+else:
+    MIXIN_BASE = object
+
+def compute_characteristic_polynomial(g, as_expression=True, bound=42):
+    """Compute the characteristic polynomial of a graph's adjacency matrix"""
+    import sympy
+    if hasattr(g, '_adjacency'):
+        A = g._adjacency
+    elif hasattr(g, '_weighted_adjacency'):
+        A = g._weighted_adjacency
+        
+    if A.shape[0] > bound:
+        Warning(f"Computing the characteristic polynomial of a graph with {A.shape[0]} nodes may take a long time")
+        
+    M = sympy.SparseMatrix(A.todense())
+    x = sympy.Symbol("x")
+    if as_expression:
+        return M.charpoly(x).as_expr()
+    return x, M.charpoly(x)
+
+
+def process_df_to_syms(ndf, as_records):
+    if as_records:
+        syms = [ai.Symbol(row) for row in ndf.to_dict('records')]
+    else:
+        syms = [ai.Symbol(str(row)) for row in ndf.values]
+    return syms
+
+def process_df_to_sym(df, as_records):
+    if as_records:
+        syms = ai.Symbol(df.to_dict('records'))
+    else:
+        syms = ai.Symbol(list(df.values))
+    return syms
+
+class Lambda(ai.Expression):
+    def __init__(self, callable: ai.Callable):
+        super().__init__()
+        def _callable(*args, **kwargs):
+            kw = {
+                'args': args,
+                'kwargs': kwargs,
+            }
+            return callable(kw)
+        self.callable: ai.Callable = _callable
+    
+    def forward(self, *args, **kwargs) -> Symbol:
+        return self.callable(*args, **kwargs)
+
+CONTEXTS = [None, 'summary', 'search', 'entities', 'relationships', 'money', 'time', 'threats']
+
+class SymbolicMixin(MIXIN_BASE):
+    
+    def __init__(self, *args, **kwargs):
+        pass
+    
+    def _encode_df_as_sym(self, context_df, as_records):
+        sym = process_df_to_sym(context_df, as_records)
+        stream = ai.Stream(Lambda(lambda x: x['args'][0]))
+        rr = ai.Symbol(list(stream(sym)))
+        return rr
+    
+    def _add_context_and_query(self, sym, query, context='summary'):
```

Comment:
Maybe factor out into a lazily loaded csv cfg?

Also, this feels like demoware, not sure we would want anyone dependent on them yet? If so, maybe we make internal for now...